corehq/apps/userreports/adapter.py

```diff
@@ -1,4 +1,5 @@
 from __future__ import absolute_import
+from __future__ import unicode_literals
```

Comment:
For some reason I recall that you should either use this in none of your code base, or all of it. Though I suppose that we already use it in some of it, so adding it file by file shouldn't be a terrible idea.

I bet @millerdev has an opinion/knowledge of this and could weigh in on it

Reply:
Ha, just brought it up over here: https://github.com/dimagi/xml2json/pull/5#pullrequestreview-73802902

Dhe drawbacks listed in [Should I import unicode_literals?](http://python-future.org/unicode_literals.html) (required reading if you want to weigh in on this topic) seem compelling too. I'm leaning against using `unicode_literals` for three main reasons:

- adding it to some files but not others is likely to cause confusion because strings will not behave the same way everywhere.
- adding it to all files in the (HQ) codebase seems infeasible at this time.
- it can introduce hard-to detect bugs in Python 2.

Reply:
Would you then push for us to decorate all strings with either `u`, `b`, or `r`, and be explicit about it?

Reply:
For `u` and `b`, yes, I think so. The downside to that is we'll have unnecessary `u` prefixes littered everywhere when we switch to Python 3, and removing them would be disruptive to blame history.

Note that `ur` is a valid prefix in Python 2 but not on Python 3, so that does make use of the `r` prefix more complicated in some cases. `r"xyz"` in Python 3 results in a Unicode literal while in Python 2 it's a byte string. `br"xyz"` results in a byte string on both. Good explanation [here](https://stackoverflow.com/a/26063975/10840).

I do like to omit the `u` prefix if the code "just works" without it on both Python 2 and 3, but that's tricky to verify, so maybe not recommended?

One more note here: I don't like that python-modernize is removing the `u` prefix on strings. It's unnecessary (we don't need to support Python 3.0-3.2) and just creates more noise in the patch.

Reply:
Another thought to entertain the other side of the argument. Maybe we should go with the [approach Django took](https://docs.djangoproject.com/en/1.11/topics/python3/). It's well articulated and I'm assuming fairly complete in addressing all the topics we'd care about. It's also a big project, so their approach is probably doable for us since they were successful with it.

One big counterargument is that they were aiming to have their code run on both Python 2 and Python 3. I'm assuming that we'll switch to Python 3 immediately and not try to maintain backward compatibility as soon as all of our code supports that to some reasonable degree of stability.

Reply:
From a process perspective, getting compatible code via ```unicode_literals``` is as simple as running:
```python-modernize corehq/apps/userreports/ -w --future-unicode --no-six```
I could imagine us rolling this out across the entire codebase in a relatively short timeframe.

I don't know of an automated way to add ```u``` in all of needed places.

Reply:
I'm not sure I agree. Check out the second item under [Drawbacks](http://python-future.org/unicode_literals.html#drawbacks) for an example.

Reply:
@dimagi/team-commcare-hq After thinking it over for a while, I'm going to make a pronouncement so we can move forward with porting to Python 3 with as little ambiguity as possible.

I propose that we use [Django's Python 3 porting guide](https://docs.djangoproject.com/en/1.11/topics/python3/).

This means we will use `unicode_literals` as described there. Since we are targeting Python ≥ 3.3, removal of `u` prefixes is optional (although still recommended). The section on [string handling](https://docs.djangoproject.com/en/1.11/topics/python3/#string-handling) is particularly important, especially in clarifying the naming switch from `str` and `unicode` to `bytes` and `text`. We should follow that convention and rename functions in our code to avoid ambiguity.

To reduce the burden of conversion and review, instead of converting the entire code base at once, we will attempt to convert entire Django apps at once (or top-level packages if not part of a Django app). Code reviewers should be vigilant for old-style string assumptions sneaking into parts of the codebase that have been ported to use `unicode_literals`.

I'm still open to being persuaded in a different direction, so if any of this seems like not a great idea, then please speak up.

Reply:
Late to the conversation, but going with django's approach sounds good. To this point:
> One big counterargument is that they were aiming to have their code run on both Python 2 and Python 3. I'm assuming that we'll switch to Python 3 immediately and not try to maintain backward compatibility as soon as all of our code supports that to some reasonable degree of stability.
Might be nitpicking, but unless it's feasible to switch to Python 3 very soon - like, within a month - I think we're better off thinking of our code as supporting both 2 and 3, and considering doing a later cleanup pass after the switch to 3. Which is to say, that counterargument doesn't seem like a big deal.